haystack/schema.py

```diff
@@ -550,36 +617,305 @@ def __setitem__(self, key: str, value: pd.DataFrame):
+        Document metrics are:
+        - mrr (Mean Reciprocal Rank: see https://en.wikipedia.org/wiki/Mean_reciprocal_rank)
+        - map (Mean Average Precision: see https://en.wikipedia.org/wiki/Evaluation_measures_(information_retrieval)#Mean_average_precision)
+        - precision (Precision: How much of the returned documents were relevant?)
+        - recall_ir (Recall according to Information Retrieval definition: How much of the relevant documents were retrieved per query?)
+        - recall_qs (Recall for Question Answering: Did the query return at least one relevant document? -> 1.0 or 0.0)
+        """
+        if simulated_top_k_retriever != -1:
+            documents = documents[documents["rank"] <= simulated_top_k_retriever]
+        
+        metrics = []
+        queries = documents["query"].unique()
+        for query in queries:
+            query_df = documents[documents["query"] == query]
+            gold_ids = query_df["gold_document_ids"].iloc[0]
+            retrieved = len(query_df)
+            
+            relevance_criteria_ids = list(query_df[query_df[doc_relevance_col] == 1]["document_id"].values)
+            num_relevants = len(set(gold_ids + relevance_criteria_ids))
+            num_retrieved_relevants = query_df[doc_relevance_col].values.sum()
+            rank_retrieved_relevants = query_df[query_df[doc_relevance_col] == 1]["rank"].values
+            avp_retrieved_relevants = [query_df[doc_relevance_col].values[:rank].sum() / rank 
+                                            for rank in rank_retrieved_relevants]
+
+            avg_precision = np.sum(avp_retrieved_relevants) / num_relevants if num_relevants > 0 else 0.0
+            recall_ir = num_retrieved_relevants / num_relevants if num_relevants > 0 else 0.0
+            recall_qa = min(num_retrieved_relevants, 1)
+            precision = num_retrieved_relevants / retrieved if retrieved > 0 else 0.0
+            rr = 1.0 / rank_retrieved_relevants.min() if len(rank_retrieved_relevants) > 0 else 0.0
+
+            metrics.append({
+                "recall_ir": recall_ir,
+                "recall_qa": recall_qa,
+                "precision": precision,
+                "map": avg_precision,
+                "mrr": rr
+            })
+
+        metrics_df = pd.DataFrame.from_records(metrics, index=queries)
+        return metrics_df
 
     def save(self, out_dir: Union[str, Path]):
+        """
+        Saves the evaluation result. 
+        The result of each node is saved in a separate csv with file name {node_name}.csv to the out_dir folder.
+
+        :param out_dir: Path to the target folder the csvs will be saved.
+        """
         out_dir = out_dir if isinstance(out_dir, Path) else Path(out_dir)
         for node_name, df in self.node_results.items():
             target_path = out_dir / f"{node_name}.csv"
             df.to_csv(target_path, index=False, header=True)
 
     @classmethod
     def load(cls, load_dir: Union[str, Path]):
```

Comment:
It's great to have this function in addition to `save()`. Do I understand correctly that the scenario when to use `load()` is for example if I want to add more evaluation metrics later on to my generated predictions?

Reply:
Yes for example. Or meanwhile there is a better way to visualize the results. Or say I made a mistake in the calculation of some metric, we can fix it and run `calculate_metrics` or `print_eval_report` again. Additionally, I think it should always be possible to reproduce the results given the raw evaluation data.